ethereum/core/src/main/java/tech/pegasys/pantheon/ethereum/mainnet/ProtocolScheduleBuilder.java

```diff
@@ -0,0 +1,96 @@
+/*
+ * Copyright 2018 ConsenSys AG.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file except in compliance with
+ * the License. You may obtain a copy of the License at
+ *
+ * http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software distributed under the License is distributed on
+ * an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the
+ * specific language governing permissions and limitations under the License.
+ */
+package tech.pegasys.pantheon.ethereum.mainnet;
+
+import tech.pegasys.pantheon.config.GenesisConfigOptions;
+
+import java.util.OptionalLong;
+import java.util.function.Function;
+
+public class ProtocolScheduleBuilder<C> {
+
+  private final GenesisConfigOptions config;
+  private final Function<ProtocolSpecBuilder<Void>, ProtocolSpecBuilder<C>> protocolSpecAdapter;
+  private final int defaultChainId;
+
+  public ProtocolScheduleBuilder(
+      final GenesisConfigOptions config,
+      final int defaultChainId,
+      final Function<ProtocolSpecBuilder<Void>, ProtocolSpecBuilder<C>> protocolSpecAdapter) {
+    this.config = config;
+    this.protocolSpecAdapter = protocolSpecAdapter;
+    this.defaultChainId = defaultChainId;
+  }
+
+  public ProtocolSchedule<C> createProtocolSchedule() {
+    final int chainId = config.getChainId().orElse(defaultChainId);
+    final MutableProtocolSchedule<C> protocolSchedule = new MutableProtocolSchedule<>(chainId);
+
+    addProtocolSpec(
+        protocolSchedule, OptionalLong.of(0), MainnetProtocolSpecs.frontierDefinition());
+    addProtocolSpec(
+        protocolSchedule,
+        config.getHomesteadBlockNumber(),
+        MainnetProtocolSpecs.homesteadDefinition());
+
+    config
+        .getDaoForkBlock()
+        .ifPresent(
+            daoBlockNumber -> {
+              if (daoBlockNumber > 0) {
```

Comment:
Slight concern that if the DAO fork occurs anywhere other than Frontier or Homestead, that the protocol spec will revert to an incorrect spec.

Given Dao fork should only occur on mainnet, and it definitively occurred in homestead - this is probably fine.

Reply:
Agreed on both counts. It's messy but looking at what the Dao fork does, chances of success on anything other than MainNet at the specific time of the Dao fork are almost certain to end in tears.